src/Interactive/EditorFeatures/Core/Extensibility/Interactive/ResetInteractive.cs

```diff
@@ -0,0 +1,134 @@
+using Microsoft.CodeAnalysis.Editor.Host;
+using Microsoft.CodeAnalysis.Editor.Interactive;
+using Microsoft.VisualStudio.InteractiveWindow;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Editor;
+
+namespace Microsoft.VisualStudio.LanguageServices.Interactive
+{
+    internal abstract class ResetInteractive
+    {
+        private readonly Func<string, string> _createReference;
+
+        private readonly Func<string, string> _createImport;
+
+        internal event EventHandler ExecutionCompleted;
+
+        internal ResetInteractive(Func<string, string> createReference, Func<string, string> createImport)
+        {
+            _createReference = createReference;
+            _createImport = createImport;
+        }
+
+        internal Task Execute(IInteractiveWindow interactiveWindow, string title)
+        {
+            List<string> references, referenceSearchPaths, sourceSearchPaths, namespacesToImport;
+            string projectDirectory;
+
+            if (GetProjectProperties(out references, out referenceSearchPaths, out sourceSearchPaths, out namespacesToImport, out projectDirectory))
+            {
+                // Now, we're going to do a bunch of async operations.  So create a wait
+                // indicator so the user knows something is happening, and also so they cancel.
+                var waitIndicator = GetWaitIndicator();
+                var waitContext = waitIndicator.StartWait(title, InteractiveEditorFeaturesResources.BuildingProject, allowCancel: true);
+
+                var resetInteractiveTask = ResetInteractiveAsync(
+                    interactiveWindow,
+                    references.ToImmutableArray(),
+                    referenceSearchPaths.ToImmutableArray(),
+                    sourceSearchPaths.ToImmutableArray(),
+                    namespacesToImport.ToImmutableArray(),
+                    projectDirectory,
+                    waitContext);
+
+                // Once we're done resetting, dismiss the wait indicator and focus the REPL window.
+                return resetInteractiveTask.SafeContinueWith(
+                    _ =>
+                    {
+                        waitContext.Dispose();
+                        ExecutionCompleted?.Invoke(this, new EventArgs());
+                    },
+                    TaskScheduler.FromCurrentSynchronizationContext());
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private async Task ResetInteractiveAsync(
+            IInteractiveWindow interactiveWindow,
+            ImmutableArray<string> referencePaths,
+            ImmutableArray<string> referenceSearchPaths,
+            ImmutableArray<string> sourceSearchPaths,
+            ImmutableArray<string> namespacesToImport,
+            string projectDirectory,
+            IWaitContext waitContext)
+        {
+            // First, open the repl window.
+            IInteractiveEvaluator engine = interactiveWindow.Evaluator;
```

Comment:
As long as we're refactoring everything, can you please rename this to "evaluator"?